sql: populate EXPLAIN(DEBUG) of UPDATE

currently:
```
explain (debug) update deletes set a = 4 where k = 6;
 RowIdx | Key | Value | Disposition
--------+-----+-------+-------------
(0 rows)
```

This is even worse now:
```
insert into system.users values ('a','b'); explain (debug) update system.users set username = 'abcd' where hashedPassword = 'b';
driver: bad connection
```
That's because the server has crashed:
```
panic: runtime error: slice bounds out of range

goroutine 772 [running]:
panic(0x52fda40, 0xc8200140c0)
	/Users/tamird/src/go1.6/src/runtime/panic.go:481 +0x3e6
github.com/cockroachdb/cockroach/sql.(*updateNode).Next(0xc8207b0900, 0xc8207b0900, 0x0, 0x0)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/update.go:296 +0x90c
github.com/cockroachdb/cockroach/sql.(*explainDebugNode).Next(0xc8207a6650, 0xc8201596c0, 0x0, 0x0)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/explain.go:381 +0x45
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt(0xc8201d2000, 0x81146f0, 0xc820cef200, 0xc82079e090, 0xc820cef201, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/executor.go:1050 +0x41c
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtInOpenTxn(0xc8201d2000, 0x81146f0, 0xc820cef200, 0xc82079e090, 0x101, 0xc82079e028, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/executor.go:908 +0x439
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtsInCurrentTxn(0xc8201d2000, 0xc820972b30, 0x1, 0x1, 0xc82079e090, 0xc82079e028, 0x5c50101, 0x0, 0x0, 0x0, ...)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/executor.go:666 +0x5da
github.com/cockroachdb/cockroach/sql.runTxnAttempt(0xc8201d2000, 0xc82079e090, 0x1, 0xc82079e028, 0xc820880520, 0xc820972b30, 0x1, 0x1, 0x0, 0x0, ...)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/executor.go:594 +0x18d
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest.func2(0xc82016a2c0, 0xc820880520, 0x0, 0x0)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/executor.go:486 +0x470
github.com/cockroachdb/cockroach/client.(*Txn).Exec(0xc82016a2c0, 0xc8207e0101, 0x145cecca2f866ee5, 0x0, 0xc8207e61e0, 0x0, 0x0)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/client/txn.go:541 +0x2eb
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc8201d2000, 0x81c0370, 0xc8207ba600, 0xc82079e000, 0xc8206a6071, 0x80, 0x0, 0x0, 0x0, 0xc8206a6000)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/executor.go:491 +0xba7
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc8201d2000, 0x81c0370, 0xc8207ba600, 0xc82079e000, 0xc8206a6071, 0x80, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/executor.go:360 +0xf6
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc82012d500, 0x81c0370, 0xc8207ba600, 0xc8206a6071, 0x80, 0x0, 0x0, 0x0, 0x0, 0x8191401, ...)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:640 +0x98
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc82012d500, 0x81c0370, 0xc8207ba600, 0xc82012d528, 0x0, 0x0)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:320 +0xe8
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc82012d500, 0x0, 0x0, 0x0)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:275 +0x100c
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc820010d80, 0x8195ef8, 0xc820178840, 0x0, 0x0)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:231 +0x9d9
github.com/cockroachdb/cockroach/server.(*Server).Start.func8.1(0x8195ef8, 0xc820178840)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/server/server.go:369 +0x42
github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith.func1(0xc8200dc140, 0x8195ef8, 0xc820178840, 0xc820492030)
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:131 +0x62
created by github.com/cockroachdb/cockroach/util/netutil.(*Server).ServeWith
	/Users/tamird/src/go/src/github.com/cockroachdb/cockroach/util/netutil/net.go:133 +0x333
```

--- a/sql/delete.go
+++ b/sql/delete.go
@@ -103,15 +103,20 @@ func (d *deleteNode) Start() error {
 		return err
 	}
 
-	// Check if we can avoid doing a round-trip to read the values and just
-	// "fast-path" skip to deleting the key ranges without reading them first.
-	// TODO(dt): We could probably be smarter when presented with an index-join,
-	// but this goes away anyway once we push-down more of SQL.
-	sel := d.run.rows.(*selectTopNode).source.(*selectNode)
-	if scan, ok := sel.source.plan.(*scanNode); ok && canDeleteWithoutScan(d.n, scan, &d.tw) {
-		d.run.fastPath = true
-		err := d.fastDelete(scan)
-		return err
+	if d.run.explain != explainDebug {
+		// Check if we can avoid doing a round-trip to read the values and just
+		// "fast-path" skip to deleting the key ranges without reading them first.
+		// TODO(dt): We could probably be smarter when presented with an index-join,
+		// but this goes away anyway once we push-down more of SQL.
+		//
+		// (When explain == explainDebug, we use the slow path so that
+		// each debugVal gets a chance to be reported via Next().)
+		sel := d.run.rows.(*selectTopNode).source.(*selectNode)
+		if scan, ok := sel.source.plan.(*scanNode); ok && canDeleteWithoutScan(d.n, scan, &d.tw) {
+			d.run.fastPath = true
+			err := d.fastDelete(scan)
+			return err
+		}
 	}
 
 	if err := d.rh.startPlans(); err != nil {
@@ -136,6 +141,10 @@ func (d *deleteNode) Next() (bool, error) {
 			err = d.tw.finalize()
 		}
 		return false, err
+	}
+
+	if d.run.explain == explainDebug {
+		return true, nil
 	}
 
 	rowVals := d.run.rows.Values()
@@ -204,6 +213,10 @@ func (d *deleteNode) Values() parser.DTuple {
 }
 
 func (d *deleteNode) MarkDebug(mode explainMode) {
+	if mode != explainDebug {
+		panic(fmt.Sprintf("unknown debug mode %d", mode))
+	}
+	d.run.explain = mode
 	d.run.rows.MarkDebug(mode)
 }
 

--- a/sql/insert.go
+++ b/sql/insert.go
@@ -258,6 +258,10 @@ func (n *insertNode) Next() (bool, error) {
 		return false, err
 	}
 
+	if n.run.explain == explainDebug {
+		return true, nil
+	}
+
 	rowVals := n.run.rows.Values()
 
 	// The values for the row may be shorter than the number of columns being
@@ -451,6 +455,10 @@ func (n *insertNode) Values() parser.DTuple {
 }
 
 func (n *insertNode) MarkDebug(mode explainMode) {
+	if mode != explainDebug {
+		panic(fmt.Sprintf("unknown debug mode %d", mode))
+	}
+	n.run.explain = mode
 	n.run.rows.MarkDebug(mode)
 }
 

--- a/sql/table_join.go
+++ b/sql/table_join.go
@@ -76,6 +76,14 @@ type joinNode struct {
 	// emptyLeft contains tuples of NULL values to use
 	// on the left for full outer joins when the filter fails.
 	emptyLeft parser.DTuple
+
+	// explain indicates whether this node is running on behalf of
+	// EXPLAIN(DEBUG).
+	explain explainMode
+
+	// doneReadingRight is used by debugNext() and DebugValues() when
+	// explain == explainDebug.
+	doneReadingRight bool
 }
 
 type joinPredicate interface {
@@ -563,6 +571,10 @@ func (n *joinNode) Ordering() orderingInfo { return n.left.Ordering() }
 
 // MarkDebug implements the planNode interface.
 func (n *joinNode) MarkDebug(mode explainMode) {
+	if mode != explainDebug {
+		panic(fmt.Sprintf("unknown debug mode %d", mode))
+	}
+	n.explain = mode
 	n.left.MarkDebug(mode)
 	n.right.MarkDebug(mode)
 }
@@ -580,23 +592,25 @@ func (n *joinNode) Start() error {
 		return err
 	}
 
-	// Load all the rows from the right side in memory.
-	v := &valuesNode{}
-	for {
-		hasRow, err := n.right.Next()
-		if err != nil {
-			return err
+	if n.explain != explainDebug {
+		// Load all the rows from the right side in memory.
+		v := &valuesNode{}
+		for {
+			hasRow, err := n.right.Next()
+			if err != nil {
+				return err
+			}
+			if !hasRow {
+				break
+			}
+			row := n.right.Values()
+			newRow := make([]parser.Datum, len(row))
+			copy(newRow, row)
+			v.rows = append(v.rows, newRow)
 		}
-		if !hasRow {
-			break
+		if len(v.rows) > 0 {
+			n.rightRows = v
 		}
-		row := n.right.Values()
-		newRow := make([]parser.Datum, len(row))
-		copy(newRow, row)
-		v.rows = append(v.rows, newRow)
-	}
-	if len(v.rows) > 0 {
-		n.rightRows = v
 	}
 
 	// Pre-allocate the space for output rows.
@@ -623,8 +637,27 @@ func (n *joinNode) Start() error {
 	return nil
 }
 
+func (n *joinNode) debugNext() (bool, error) {
+	if !n.doneReadingRight {
+		hasRightRow, err := n.right.Next()
+		if err != nil {
+			return false, err
+		}
+		if hasRightRow {
+			return true, nil
+		}
+		n.doneReadingRight = true
+	}
+
+	return n.left.Next()
+}
+
 // Next implements the planNode interface.
 func (n *joinNode) Next() (bool, error) {
+	if n.explain == explainDebug {
+		return n.debugNext()
+	}
+
 	var leftRow, rightRow parser.DTuple
 	var nRightRows int
 
@@ -733,7 +766,14 @@ func (n *joinNode) Values() parser.DTuple {
 
 // DebugValues implements the planNode interface.
 func (n *joinNode) DebugValues() debugValues {
-	// TODO(knz): not so clear yet how to report DebugValues()
-	// for JOIN.
-	return debugValues{}
+	var res debugValues
+	if !n.doneReadingRight {
+		res = n.right.DebugValues()
+	} else {
+		res = n.left.DebugValues()
+	}
+	if res.output == debugValueRow {
+		res.output = debugValueBuffered
+	}
+	return res
 }

--- a/sql/testdata/explain_debug
+++ b/sql/testdata/explain_debug
@@ -209,3 +209,65 @@ EXPLAIN (DEBUG) SELECT * FROM ab WHERE a > 1 INTERSECT SELECT * FROM ab WHERE b 
 1  /ab/primary/2/2  NULL  ROW
 2  /ab/primary/2/6  NULL  ROW
 3  /ab/primary/3/9  NULL  ROW
+
+query ITTT
+EXPLAIN (DEBUG) INSERT INTO ab(a, b) VALUES (42, 51)
+----
+0 0 (42, 51) ROW
+
+query ITTT
+EXPLAIN (DEBUG) INSERT INTO ab(a, b) SELECT a+1, b+1 FROM ab WHERE b > 6
+----
+0  /ab/primary/1/4  NULL  FILTERED
+1  /ab/primary/1/5  NULL  FILTERED
+2  /ab/primary/2/1  NULL  FILTERED
+3  /ab/primary/2/2  NULL  FILTERED
+4  /ab/primary/2/6  NULL  FILTERED
+5  /ab/primary/3/9  NULL  ROW
+
+query ITTT
+EXPLAIN (DEBUG) UPDATE ab SET a = a + 1 WHERE b > 6
+----
+0  /ab/primary/1/4  NULL  FILTERED
+1  /ab/primary/1/5  NULL  FILTERED
+2  /ab/primary/2/1  NULL  FILTERED
+3  /ab/primary/2/2  NULL  FILTERED
+4  /ab/primary/2/6  NULL  FILTERED
+5  /ab/primary/3/9  NULL  ROW
+
+query ITTT
+EXPLAIN (DEBUG) DELETE FROM ab WHERE b > 6
+----
+0  /ab/primary/1/4  NULL  FILTERED
+1  /ab/primary/1/5  NULL  FILTERED
+2  /ab/primary/2/1  NULL  FILTERED
+3  /ab/primary/2/2  NULL  FILTERED
+4  /ab/primary/2/6  NULL  FILTERED
+5  /ab/primary/3/9  NULL  ROW
+
+query ITTT
+EXPLAIN (DEBUG) DELETE FROM ab
+----
+0  /ab/primary/1/4  NULL  ROW
+1  /ab/primary/1/5  NULL  ROW
+2  /ab/primary/2/1  NULL  ROW
+3  /ab/primary/2/2  NULL  ROW
+4  /ab/primary/2/6  NULL  ROW
+5  /ab/primary/3/9  NULL  ROW
+
+statement ok
+UPDATE abc SET c = 1.1 WHERE a = 1
+
+query ITTT
+EXPLAIN (DEBUG) SELECT * FROM ab, abc WHERE abc.a+ab.b = abc.a*ab.a
+----
+0  /abc/primary/1/'one'    NULL  PARTIAL
+0  /abc/primary/1/'one'/c  1.1   BUFFERED
+1  /abc/primary/2/'two'    NULL  BUFFERED
+2  /abc/primary/3/'three'  NULL  BUFFERED
+0  /ab/primary/1/4         NULL  BUFFERED
+1  /ab/primary/1/5         NULL  BUFFERED
+2  /ab/primary/2/1         NULL  BUFFERED
+3  /ab/primary/2/2         NULL  BUFFERED
+4  /ab/primary/2/6         NULL  BUFFERED
+5  /ab/primary/3/9         NULL  BUFFERED

--- a/sql/update.go
+++ b/sql/update.go
@@ -62,6 +62,8 @@ type editNodeRun struct {
 	rows      planNode
 	tw        tableWriter
 	resultRow parser.DTuple
+
+	explain explainMode
 }
 
 func (r *editNodeRun) initEditNode(en *editNodeBase, rows planNode, re parser.ReturningExprs, desiredTypes []parser.Datum) error {
@@ -287,6 +289,10 @@ func (u *updateNode) Next() (bool, error) {
 		return false, err
 	}
 
+	if u.run.explain == explainDebug {
+		return true, nil
+	}
+
 	tracing.AnnotateTrace()
 
 	oldValues := u.run.rows.Values()
@@ -379,6 +385,10 @@ func (u *updateNode) Values() parser.DTuple {
 }
 
 func (u *updateNode) MarkDebug(mode explainMode) {
+	if mode != explainDebug {
+		panic(fmt.Sprintf("unknown debug mode %d", mode))
+	}
+	u.run.explain = mode
 	u.run.rows.MarkDebug(mode)
 }
 

